public/app/features/alerting/unified/hooks/ruleGroup/useUpsertRuleFromRuleGroup.ts

```diff
@@ -0,0 +1,146 @@
+import { produce } from 'immer';
+import { isEqual } from 'lodash';
+
+import { t } from 'app/core/internationalization';
+import { dispatch } from 'app/store/store';
+import { RuleGroupIdentifier, EditableRuleIdentifier } from 'app/types/unified-alerting';
+import { PostableRuleDTO } from 'app/types/unified-alerting-dto';
+
+import { alertRuleApi } from '../../api/alertRuleApi';
+import { addRuleAction, updateRuleAction } from '../../reducers/ruler/ruleGroups';
+import { fetchRulerRulesAction } from '../../state/actions';
+import { isGrafanaRuleIdentifier, isGrafanaRulerRule } from '../../utils/rules';
+import { useAsync } from '../useAsync';
+
+import { useDeleteRuleFromGroup } from './useDeleteRuleFromGroup';
+import { useProduceNewRuleGroup } from './useProduceNewRuleGroup';
+
+/**
+ * This hook will add a single rule to a rule group – a new rule group will be created if it does not already exist.
+ */
+export function useAddRuleToRuleGroup() {
+  const [produceNewRuleGroup] = useProduceNewRuleGroup();
+  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();
+
+  const successMessage = t('alerting.rules.add-rule.success', 'Rule added successfully');
+
+  return useAsync(async (ruleGroup: RuleGroupIdentifier, rule: PostableRuleDTO, interval?: string) => {
+    const { namespaceName, dataSourceName } = ruleGroup;
+
+    // the new rule might have to be created in a new group, pass name and interval (optional) to the action
+    const action = addRuleAction({ rule, interval, groupName: ruleGroup.groupName });
+    const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);
+
+    const result = upsertRuleGroup({
+      rulerConfig,
+      namespace: namespaceName,
+      payload: newRuleGroupDefinition,
+      notificationOptions: { successMessage },
+    }).unwrap();
+
+    // @TODO remove
+    await dispatch(fetchRulerRulesAction({ rulesSourceName: dataSourceName }));
```

Comment:
We should definitely `await` the result here, the reason this works is because refreshing the rules takes longer than upserting (when we call `upsertRuleGroup` the Promise is created and the requests is sent whether we are awaiting the result or not) so the order would still be

1. send request to upsert
2. send refresh request
3. do whatever else is awaiting the result of `useUpdateRuleInRuleGroup()`